internal/match/resyntax.go

```diff
@@ -0,0 +1,491 @@
+	} else {
+		text = fmt.Sprintf("syntax error at offset %d", e.Offset)
+	}
+	if e.Context != "" {
+		text += " near `" + e.Context + "`"
+	}
+	text += ": " + e.Err
+	return text
+}
+
+// reSyntaxError returns a *SyntaxError with context.
+func reSyntaxError(s string, i int, err error) error {
+	context := s[:i]
+	j := i - 10
+	for j >= 0 && s[j] != ' ' {
+		j--
+	}
+	if j >= 0 && i-j < 20 {
+		context = s[j+1 : i]
+	} else {
+		if j < 0 {
+			j = 0
+		}
+		context = s[j:i]
+	}
+	return &SyntaxError{
+		Offset:  i,
+		Context: context,
+		Err:     err.Error(),
+	}
+}
+
+// A reSyntax is a regexp syntax tree.
+type reSyntax struct {
+	op  reOp        // opcode
+	sub []*reSyntax // subexpressions
+	w   []WordID    // words
+	n   int32       // wildcard count or match result
+}
+
+// A reOp is the opcode for a regexp syntax tree node.
+type reOp int
+
+const (
+	opNone reOp = 1 + iota
+	opEmpty
+	opWords
+	opConcat
+	opAlternate
+	opWild
+	opQuest
+
+	// pseudo-ops during parsing
+	opPseudo
+	opLeftParen
+	opVerticalBar
+)
+
+// string returns a text form for the regexp syntax.
+// The dictionary d supplies the word literals.
+func (re *reSyntax) string(d *Dict) string {
+	var b bytes.Buffer
+	rePrint(&b, re, d)
+	return strings.Trim(b.String(), "\n")
+}
+
+// nl appends a \n to b, removing any end-of-line trailing spaces and blank lines.
+func nl(b *bytes.Buffer) {
+	buf := b.Bytes()
+	if len(buf) == 0 || buf[len(buf)-1] == '\n' {
+		return
+	}
+	i := len(buf)
+	for i > 0 && buf[i-1] == ' ' {
+		i--
+	}
+	if i < len(buf) {
+		b.Truncate(i)
+	}
+	b.WriteByte('\n')
+}
+
+// rePrint prints re to b, using d for words.
+func rePrint(b *bytes.Buffer, re *reSyntax, d *Dict) {
+	switch re.op {
+	case opEmpty:
+		b.WriteString("(( ))")
+	case opConcat:
+		if len(re.sub) == 0 {
+			b.WriteString("«empty concat»")
+		}
+		for i, sub := range re.sub {
+			if i > 0 && b.Len() > 0 && b.Bytes()[b.Len()-1] != '\n' {
+				b.WriteString(" ")
+			}
+			rePrint(b, sub, d)
+		}
+
+	case opAlternate:
+		nl(b)
+		b.WriteString("((")
+		for i, sub := range re.sub {
+			if i > 0 {
+				b.WriteString(" || ")
+			}
+			rePrint(b, sub, d)
+		}
+		b.WriteString("))\n")
+
+	case opWild:
+		fmt.Fprintf(b, "__%d__", re.n)
+
+	case opQuest:
+		sub := re.sub[0]
+		nl(b)
+		if sub.op == opAlternate {
+			rePrint(b, sub, d)
+			b.Truncate(b.Len() - 1) // strip \n
+		} else {
+			b.WriteString("((")
+			rePrint(b, sub, d)
+			b.WriteString("))")
+		}
+		b.WriteString("??\n")
+
+	case opWords:
+		if len(re.w) == 0 {
+			b.WriteString("«empty opWords»")
+		}
+		for i, w := range re.w {
+			if i > 0 && b.Len() > 0 && b.Bytes()[b.Len()-1] != '\n' {
+				b.WriteString(" ")
+			}
+			s := d.Words()[w]
+			if s == "" {
+				b.WriteString("''")
+			} else {
+				b.WriteString(d.Words()[w])
+			}
+		}
+	}
+}
+
+// A reParser is the regexp parser state.
+type reParser struct {
+	dict  *Dict
+	stack []*reSyntax
+}
+
+// reParse parses a license regexp s
```

Comment:
OOC: Why this "blank verse" style of comment where you break lines at clauses, rather than standard wrapping?

Reply:
https://rhodesmill.org/brandon/2012/one-sentence-per-line/